packages/interactive_media_ads/lib/src/android/android_ad_display_container.dart

```diff
@@ -0,0 +1,241 @@
+
+  // Handles ad playback.
+  late final ima.VideoView _videoView;
+  ima.MediaPlayer? _mediaPlayer;
+
+  /// Callbacks that update the state of ad playback.
+  @internal
+  final Set<ima.VideoAdPlayerCallback> videoAdPlayerCallbacks =
+      <ima.VideoAdPlayerCallback>{};
+
+  // Handles ad playback callbacks from the IMA SDK.
+  late final ima.VideoAdPlayer _videoAdPlayer;
+
+  /// The native Android AdDisplayContainer.
+  @internal
+  ima.AdDisplayContainer? adDisplayContainer;
+
+  // Currently loaded ad.
+  ima.AdMediaInfo? _loadedAdMediaInfo;
+
+  // The saved ad position, used to resume ad playback following an ad
+  // click-through.
+  int _savedAdPosition = 0;
+
+  // Timer used to periodically update the IMA SDK of the progress of the
+  // currently playing ad.
+  Timer? _adProgressTimer;
+
+  int? _adDuration;
+
+  late final AndroidAdDisplayContainerCreationParams _androidParams =
+      params is AndroidAdDisplayContainerCreationParams
+          ? params as AndroidAdDisplayContainerCreationParams
+          : AndroidAdDisplayContainerCreationParams
+              .fromPlatformAdDisplayContainerCreationParams(params);
+
+  @override
+  Widget build(BuildContext context) {
+    return AndroidViewWidget(
+      key: params.key,
+      view: _frameLayout,
+      platformViewsServiceProxy: _androidParams._platformViewsProxy,
+      onPlatformViewCreated: () async {
+        adDisplayContainer = await _androidParams._imaProxy
+            .createAdDisplayContainerImaSdkFactory(
+          _frameLayout,
+          _videoAdPlayer,
+        );
+        params.onContainerAdded(this);
+      },
+    );
+  }
+
+  void _resetPlayer() {
+    _mediaPlayer = null;
+    _savedAdPosition = 0;
+  }
+
+  // Starts periodically updating the IMA SDK the progress of the currently
+  // playing ad.
+  void _startAdProgressTracking() {
+    // Stop any previous ad tracking.
+    _stopAdProgressTracking();
+    _adProgressTimer = Timer.periodic(
+      const Duration(milliseconds: _progressPollingMs),
+      (Timer timer) async {
+        final ima.VideoProgressUpdate currentProgress =
+            _androidParams._imaProxy.newVideoProgressUpdate(
+          currentTimeMs: await _videoView.getCurrentPosition(),
+          durationMs: _adDuration!,
+        );
+        await Future.wait(
+          <Future<void>>[
+            _videoAdPlayer.setAdProgress(currentProgress),
+            ...videoAdPlayerCallbacks.map(
+              (ima.VideoAdPlayerCallback callback) async {
+                await callback.onAdProgress(
+                  _loadedAdMediaInfo!,
+                  currentProgress,
+                );
+              },
+            ),
+          ],
+        );
+      },
+    );
+  }
+
+  // Stops updating the IMA SDK the progress of the currently playing ad.
+  void _stopAdProgressTracking() {
+    _adProgressTimer?.cancel();
+    _adProgressTimer = null;
+  }
+
+  static ima.VideoView _setUpVideoView(
```

Comment:
I started a new style of setting callback methods, which can cause a circular reference, in a static method. I would forget the rule of not using non-local methods in callback methods and this help prevented me from accidentally calling a method/field in the encapsulating class.

Reply:
It's probably worth briefly commenting this pattern in the code each time it's used so that future people understand the benefits and aren't tempted to "simplify" the code by removing this (or at least not replicating it in new code).